src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java

```diff
@@ -0,0 +1,96 @@
+package org.jabref.logic.importer.fetcher;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLConnection;
+import java.util.Objects;
+import java.util.Optional;
+
+import org.jabref.logic.importer.FulltextFetcher;
+import org.jabref.model.entry.BibEntry;
+
+import kong.unirest.Unirest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * FulltextFetcher implementation that attempts to find a PDF URL at APS. Also see the <a
+ * href="https://harvest.aps.org/docs/harvest-api">API</a>, although it isn't currently used.
+ */
+public class ApsFetcher implements FulltextFetcher {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ApsFetcher.class);
+
+    // The actual API needs either an API key or a header. This is a workaround.
+    private static final String DOI_URL = "https://www.doi.org/";
+    private static final String PDF_URL = "https://journals.aps.org/prl/pdf/";
+
+    @Override
+    public Optional<URL> findFullText(BibEntry entry) throws IOException {
+        Objects.requireNonNull(entry);
+
+        return entry.getDOI().map(doi -> {
```

Comment:
The code is more complicated due to functional style (which needs to catch exceptions etc). Thus either use a normal version (with quick exit ala `if(entry.getDoi().isEmpty()) return empty() else ... `) or alternatively try to reduce the exceptions thrown. For example, the FileNotFoundEx is only thrown to signal the the id is empty (if I'm not mistaken). A prefered option would be to make `getId` return `optional<string>` and than use `flatMap` instead of `map`.

In summary, the functional style is ok if the code could be refactored such that the following works:
````
return entry.getDOI().flatMap(this::determineId).flatMap(this::determinePdfUrl);

```